external_services/lti.py

```diff
@@ -126,26 +136,32 @@ def sign_get_query(self, url=None):
 
 class CustomStudentInfoLTIRequest(LTIRequest):
 
-    def __init__(self, service, user, profiles, instance, host, title, context_id=None, link_id=None, add=None):
+    def __init__(self, service, user, profiles, instance, request, title, context_id=None, link_id=None, add=None, exercise=None):
+        self.service = service
+        self.course_instance = instance
         parameters = add or {}
-        parameters['custom_student_id'] = self.true_student_id(user.userprofile)
+        if not service.is_anonymous:
+            parameters['custom_student_id'] = self.true_student_id(user.userprofile)
         if len(profiles) > 1:
             parameters['custom_group_members'] = self.group_json(profiles)
-        super().__init__(service, user, instance, host, title, context_id, link_id, parameters)
+        super().__init__(service, user, instance, request, title, context_id, link_id, parameters, exercise)
 
     def true_student_id(self, profile):
         return profile.student_id or "A{:d}".format(profile.id)
 
     def group_json(self, profiles):
-        data = [];
+        data = []
         for profile in profiles:
             user = profile.user
-            data.append({
-                'user': self.external_student_id(user),
-                'student_id': self.true_student_id(profile),
-                'given_name': user.first_name,
-                'family_name': user.last_name,
-                'full_name': "{} {}".format(user.first_name, user.last_name),
-                'email': user.email,
-            })
+            student_id, given_name, family_name, full_name, email = self.user_info(self.course_instance, user)
+            d = {
+                'user': student_id,
+                'given_name': given_name,
+                'family_name': family_name,
+                'full_name': full_name,
+                'email': email,
+            }
+            if not self.service.is_anonymous:
+                d['student_id'] = self.true_student_id(profile)
```

Comment:
This seems like a hack for services that expect real student id instead of internal reference. I think we should check what services use this (I presume rubyric is the only one) and we should change this as it's not consistent with other places.  I think we should deliver real student id in differently named field if possible (and thus we wouldn't over write aplus student id).

Reply:
The `if not self.service.is_anonymous:` part is not modifying any existing key in the dictionary. It adds a new key, but only if the service is not anonymous since it is sensitive user data.

Reply:
Damn. I got fooled my self as this has the student_id with different name (user).

It actually would be good idea to rename `student_id` to `user_id` in the code as it's more describing. We are not actually sending student ID's to external services (except here and another custom field).

Reply:
I have renamed `student_id` to `user_id`.